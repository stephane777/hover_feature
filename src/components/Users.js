import React from "react";
import UserList from "../components/UserList";
import UserCard from "../components/UserCard";
import PropTypes from "prop-types";

const userList = (user, dispatch) => {
	const props = {
		key: user.login.uuid,
		data: user,
		dispatch,
	};
	return <UserList {...props} />;
};

const userCard = (user) => {
	const props = {
		key: user.login.uuid,
		data: user,
	};
	return <UserCard {...props} />;
};

const Users = ({ users, route, dispatch }) => {
	return users.map((user) => {
		return route === "userList" ? userList(user, dispatch) : userCard(user);
	});
};

Users.propTypes = {
	users: PropTypes.array.isRequired,
	route: PropTypes.oneOf(["userList", "userCard"]),
};

Users.defaultProps = {
	route: "userList",
};
export default Users;
