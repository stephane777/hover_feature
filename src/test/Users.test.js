import React from "react";
import Users from "../components/Users";
import { render, screen, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import renderer, { act } from "react-test-renderer";

import { twoUsers } from "./utils";

const dispatch = jest.fn();

describe.only("test User component", () => {
	it("should render the User component with default userList props", () => {
		let mainContainer;
		expect.assertions(1);
		act(() => {
			const { container } = render(
				<Users users={twoUsers} dispatch={dispatch} />
			);
			mainContainer = container;
		});
		expect(mainContainer.firstChild.className).toBe("userList");
		// expect()
	});
	it("should show an image, title, first and lastname", () => {
		expect.assertions(23);
		act(() => {
			render(<Users users={twoUsers} dispatch={dispatch} />);
		});
		expect(screen.getAllByRole("img").length).toBe(2);
		twoUsers.forEach(
			({
				login: { uuid, username },
				name: { title, first, last },
				picture: { thumbnail },
				dob: { age },
				email,
				phone,
				location: {
					street: { number, name },
					postcode,
					city,
				},
			}) => {
				const userContainer = document.getElementById(uuid);
				const utils = within(userContainer);
				const img = utils.getByRole("img");
				expect(img).toHaveAttribute("src", thumbnail);
				expect(img).toHaveAttribute("alt", `${title} ${first} ${last}`);
				expect(utils.getByText(title)).toBeInTheDocument();
				expect(utils.getByText(first)).toBeInTheDocument();
				expect(utils.getByText(last)).toBeInTheDocument();

				expect(utils.queryByText(age)).toBeNull();
				expect(utils.queryByText(email)).toBeNull();
				expect(utils.queryByText(phone)).toBeNull();
				expect(utils.queryByText(`${number} ${name}`)).toBeNull();
				expect(utils.queryByText(postcode)).toBeNull();
				expect(utils.queryByText(city)).toBeNull();
			}
		);
	});

	it("Should render user card: ", () => {
		const { container } = render(
			<Users users={twoUsers} route="userCard" dispatch={dispatch} />
		);
		expect(container.firstChild.className).toBe("userCard");
	});
	it("should render all user details", () => {
		// expect.assertions(16);
		act(() => {
			render(<Users users={twoUsers} route="userCard" dispatch={dispatch} />);
		});
		twoUsers.forEach(
			({
				login: { uuid, username },
				name: { title, first, last },
				picture: { thumbnail },
				dob: { age },
				email,
				phone,
				cell,
				location: {
					street: { number, name },
					postcode,
					city,
				},
			}) => {
				const userContainer = document.getElementById(uuid);
				const utils = within(userContainer);
				const img = utils.getByRole("img");
				expect(img).toHaveAttribute("src", thumbnail);
				expect(img).toHaveAttribute("alt", `${title} ${first} ${last}`);
				expect(
					screen.getByText(`${title} ${first} ${last} ${age}`)
				).toBeInTheDocument();

				expect(screen.getByText(username));
				expect(screen.getByText(email)).toBeInTheDocument();
				expect(screen.getByText(phone)).toBeInTheDocument();
				expect(screen.getByText(cell)).toBeInTheDocument();
				expect(screen.getAllByText(/Load more/)).toHaveLength(2);
			}
		);
	});

	describe("Snapshot testing", () => {
		let root;
		it("should match snapshot with userList", async () => {
			act(() => {
				root = renderer.create(<Users users={twoUsers} dispatch={dispatch} />);
			});
			expect(await root.toJSON()).toMatchSnapshot();
		});
		it("should match snapshot with userCard", async () => {
			act(() => {
				root.update(
					<Users users={twoUsers} route="userCard" dispatch={dispatch} />
				);
			});
			expect(await root.toJSON()).toMatchSnapshot();
		});
	});
});
