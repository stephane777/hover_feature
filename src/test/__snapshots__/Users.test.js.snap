// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test User component Snapshot testing should match snapshot with userCard 1`] = `
Array [
  <div
    className="userCard"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      id="488edda9-f042-434f-8162-e4dddf7055f5"
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "column",
          "fontSize": "1.4rem",
          "height": "100%",
          "justifyContent": "space-around",
          "overflow": "hidden",
          "width": "100%",
        }
      }
    >
      <div
        className="userCard__header"
      >
        <div
          className="userCard__header_img"
        >
          <img
            alt="Mrs Emily Mortensen"
            src="https://randomuser.me/api/portraits/thumb/women/78.jpg"
          />
        </div>
        <div
          className="userCard__header_details"
        >
          <div>
            <span
              className="userCard__title"
            >
              Mrs Emily Mortensen 31
            </span>
          </div>
          <div
            className="bold"
          >
            <span>
              blackwolf380
            </span>
          </div>
          <div>
            <span>
              emily.mortensen@example.com
            </span>
          </div>
          <div>
            <span>
              94033936
            </span>
          </div>
          <div>
            <span>
              92381351
            </span>
          </div>
        </div>
      </div>
      <div
        className="userCard__more"
      >
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "fontWeight": "bold",
            }
          }
          type="button"
        >
          Load more
        </button>
      </div>
    </div>
  </div>,
  <div
    className="userCard"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      id="bdc71f4e-e874-4575-96b5-5cfc6959458e"
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "column",
          "fontSize": "1.4rem",
          "height": "100%",
          "justifyContent": "space-around",
          "overflow": "hidden",
          "width": "100%",
        }
      }
    >
      <div
        className="userCard__header"
      >
        <div
          className="userCard__header_img"
        >
          <img
            alt="Miss Nurdan Tazeg端l"
            src="https://randomuser.me/api/portraits/thumb/women/9.jpg"
          />
        </div>
        <div
          className="userCard__header_details"
        >
          <div>
            <span
              className="userCard__title"
            >
              Miss Nurdan Tazeg端l 31
            </span>
          </div>
          <div
            className="bold"
          >
            <span>
              bigdog994
            </span>
          </div>
          <div>
            <span>
              nurdan.tazegul@example.com
            </span>
          </div>
          <div>
            <span>
              (012)-335-1433
            </span>
          </div>
          <div>
            <span>
              (010)-927-6700
            </span>
          </div>
        </div>
      </div>
      <div
        className="userCard__more"
      >
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "fontWeight": "bold",
            }
          }
          type="button"
        >
          Load more
        </button>
      </div>
    </div>
  </div>,
]
`;

exports[`test User component Snapshot testing should match snapshot with userList 1`] = `
Array [
  <div
    className="userList"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      id="488edda9-f042-434f-8162-e4dddf7055f5"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <img
        alt="Mrs Emily Mortensen"
        src="https://randomuser.me/api/portraits/thumb/women/78.jpg"
      />
      <span
        className="userList__title"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Mrs
      </span>
      <span
        className="userList__first"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Emily
      </span>
      <span
        className="userList__last"
        style={
          Object {
            "fontSize": "12px",
            "marginRight": "auto",
          }
        }
      >
        Mortensen
      </span>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Edit
      </button>
    </div>
  </div>,
  <div
    className="userList"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      id="bdc71f4e-e874-4575-96b5-5cfc6959458e"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <img
        alt="Miss Nurdan Tazeg端l"
        src="https://randomuser.me/api/portraits/thumb/women/9.jpg"
      />
      <span
        className="userList__title"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Miss
      </span>
      <span
        className="userList__first"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      >
        Nurdan
      </span>
      <span
        className="userList__last"
        style={
          Object {
            "fontSize": "12px",
            "marginRight": "auto",
          }
        }
      >
        Tazeg端l
      </span>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Edit
      </button>
    </div>
  </div>,
]
`;
